File: lintcode/Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: A list of lists of integer include 
     *          the zigzag level order traversal of its nodes' values 
     */
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root) {
        // write your code here
        if(!root) return {};
        
        queue<TreeNode *> p, q;
        p.push(root);
        
        vector<vector<int>> result;
        vector<int> layer;
        bool flag = true;
        
        while(!p.empty()){
            TreeNode *t = p.front();
            p.pop();
            
            layer.push_back(t->val);
            if(t->left) q.push(t->left);
            if(t->right) q.push(t->right);
            
            if(p.empty()){
                swap(p, q);

                if(flag == false) reverse(layer.begin(), layer.end());
                result.push_back(layer);
                layer.clear();
                flag = !flag;
            }
        }
        
        return result;
    }
};

File: leetcode/java/BinaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null) return Collections.emptyList();

        // switch between two stacks
        Deque<TreeNode> p = new ArrayDeque<>();
        p.push(root);
        
        Deque<TreeNode> q = new ArrayDeque<>();
        boolean flag = true;
        List<Integer> list = new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();

        while(!p.isEmpty()) {
            TreeNode t = p.pop();
            list.add(t.val);
            pushToStack(t, q, flag);
            if(p.isEmpty()) {
                result.add(new ArrayList<>(list));
                list.clear(); 
                flag = !flag;
                Deque<TreeNode> k = p;
                p = q;
                q = k;
            }
        }
        return result;
    }
    
    private void pushToStack(TreeNode t, Deque<TreeNode> q, boolean flag) {
        if(flag) {
            if(t.left != null) q.push(t.left);
            if(t.right != null) q.push(t.right);
        } else {
            if(t.right != null) q.push(t.right);
            if(t.left != null) q.push(t.left);
        }
    }
}
